server/src/internalClusterTest/java/org/opensearch/cluster/coordination/NodeJoinLeftIT.java

```diff
@@ -0,0 +1,368 @@
+            logger.info("--> re-enabling normal followerchecker and validating cluster is stable");
+            succeedFollowerChecker.set(true);
+            ClusterHealthResponse response2 = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+            assertThat(response2.isTimedOut(), is(false));
+
+            Thread.sleep(1000);
+            // checking again to validate stability and ensure node did not leave
+            ClusterHealthResponse response3 = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+            assertThat(response3.isTimedOut(), is(false));
+        }
+
+        succeedFollowerChecker.set(true);
+        response = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+        assertThat(response.isTimedOut(), is(false));
+    }
+
+    public void testClusterStabilityWhenDisconnectDuringSlowNodeLeftTask() throws Exception {
+        final String indexName = "test";
+        final Settings nodeSettings = Settings.builder()
+            .put(RecoverySettings.INDICES_RECOVERY_RETRY_DELAY_NETWORK_SETTING.getKey(), "100ms")
+            .put(NodeConnectionsService.CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.getKey(), "10s")
+            .put(FollowersChecker.FOLLOWER_CHECK_TIMEOUT_SETTING.getKey(), "200ms")
+            .put(FollowersChecker.FOLLOWER_CHECK_INTERVAL_SETTING.getKey(), "100ms")
+            .put(FollowersChecker.FOLLOWER_CHECK_RETRY_COUNT_SETTING.getKey(), 1)
+            .put(NodeConnectionsService.CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.getKey(), "100ms")
+            .build();
+        // start a 3 node cluster with 1 cluster-manager
+        final String cm = internalCluster().startNode(nodeSettings);
+        internalCluster().startNode(Settings.builder().put("node.attr.color", "blue").put(nodeSettings).build());
+        final String redNodeName = internalCluster().startNode(Settings.builder().put("node.attr.color", "red").put(nodeSettings).build());
+
+        ClusterHealthResponse response = client().admin().cluster().prepareHealth().setWaitForNodes(">=3").get();
+        assertThat(response.isTimedOut(), is(false));
+
+        client().admin()
+            .indices()
+            .prepareCreate(indexName)
+            .setSettings(
+                Settings.builder()
+                    .put(IndexMetadata.INDEX_ROUTING_INCLUDE_GROUP_SETTING.getKey() + "color", "blue")
+                    .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                    .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+            )
+            .get();
+
+        ClusterService cmClsService = internalCluster().getInstance(ClusterService.class, cm);
+        // Simulate a slow applier on the cm to delay node-left state application
+        cmClsService.addStateApplier(event -> {
+            if (event.nodesRemoved()) {
+                try {
+                    Thread.sleep(3000);
```

Comment:
is this sufficient to generate the race condition, do this test fail without your fix?

Reply:
yes, the test fails. validated that the same exception in FollowersChecker - `NodeNotConnectedException` is showing up